File: aoc/2025/11/part2/solver2.py
from __future__ import annotations

import sys


class Hashable:
    def hash_code(self) -> int:
        raise NotImplementedError()


class Equalable:
    def equals(self, other: object) -> bool:
        raise NotImplementedError()


class MyMap:
    def __init__(self, capacity: int, keys: list[object | None], values: list[object | None], occupied: list[bool]) -> None:
        self._len = 0
        self._capacity = capacity
        self._keys = keys
        self._values = values
        self._occupied = occupied

    @staticmethod
    def new() -> MyMap:
        return MyMap.with_capacity(8)

    @staticmethod
    def with_capacity(min_capacity: int) -> MyMap:
        capacity = 8
        while capacity < min_capacity:
            capacity = capacity * 2
        return MyMap(capacity, [None] * capacity, [None] * capacity, [False] * capacity)

    def len(self) -> int:
        return self._len

    def contains(self, key: object) -> bool:
        return self._find_existing_index(key) >= 0

    def index_get(self, key: object) -> object:
        index = self._find_existing_index(key)
        if index < 0:
            raise RuntimeError("Map.index_get: key not found")
        return self._values[index]

    def put(self, key: object, value: object) -> None:
        self._maybe_grow_for_insert()
        if self._insert_or_assign(key, value):
            self._len = self._len + 1

    def _maybe_grow_for_insert(self) -> None:
        if ((self._len + 1) * 10) < (self._capacity * 7):
            return
        self._rehash(self._capacity * 2)

    def _rehash(self, new_capacity: int) -> None:
        old_capacity = self._capacity
        old_keys = self._keys
        old_values = self._values
        old_occupied = self._occupied

        self._capacity = new_capacity
        self._keys = [None] * new_capacity
        self._values = [None] * new_capacity
        self._occupied = [False] * new_capacity
        self._len = 0

        i = 0
        while i < old_capacity:
            if old_occupied[i]:
                if self._insert_or_assign(old_keys[i], old_values[i]):
                    self._len = self._len + 1
            i = i + 1

    def _find_existing_index(self, key: object) -> int:
        capacity_i64 = self._capacity
        index = index_for(key, self._capacity)
        probes = 0

        while probes < capacity_i64:
            if not self._occupied[index]:
                return -1
            if as_equalable(self._keys[index]).equals(key):
                return index
            index = index + 1
            if index >= capacity_i64:
                index = 0
            probes = probes + 1

        return -1

    def _insert_or_assign(self, key: object, value: object) -> bool:
        capacity_i64 = self._capacity
        index = index_for(key, self._capacity)
        probes = 0

        while probes < capacity_i64:
            if not self._occupied[index]:
                self._occupied[index] = True
                self._keys[index] = key
                self._values[index] = value
                return True
            if as_equalable(self._keys[index]).equals(key):
                self._values[index] = value
                return False
            index = index + 1
            if index >= capacity_i64:
                index = 0
            probes = probes + 1

        raise RuntimeError("Map: table full")


class MyStr(Hashable, Equalable):
    def __init__(self, raw: str) -> None:
        self.raw = raw

    def hash_code(self) -> int:
        hash_value = 14695981039346656037
        for byte in self.raw.encode("utf-8"):
            hash_value = hash_value ^ byte
            hash_value = hash_value * 1099511628211
        return hash_value

    def equals(self, other: object) -> bool:
        if other is None:
            return False
        if not isinstance(other, MyStr):
            return False
        return self.raw == other.raw


class CacheElem(Hashable, Equalable):
    def __init__(self, start: int, end: int) -> None:
        self.start = start
        self.end = end

    def hash_code(self) -> int:
        hash_value = 117
        hash_value = hash_value * 311 + self.start
        hash_value = hash_value * 311 + self.end
        return hash_value

    def equals(self, other: object) -> bool:
        if other is None:
            return False
        if not isinstance(other, CacheElem):
            return False
        return self.start == other.start and self.end == other.end


def find_all_paths_dag(graph: list[list[int]], memoize_cache: MyMap, start: int, end: int) -> int:
    if start == end:
        return 1

    elem = CacheElem(start, end)
    if memoize_cache.contains(elem):
        return int(memoize_cache.index_get(elem))

    paths = 0
    for node in graph[start]:
        paths = paths + find_all_paths_dag(graph, memoize_cache, node, end)

    memoize_cache.put(elem, paths)
    return paths


def main() -> None:
    input_text = read_stdin()
    total = 0
    for i in range(0, 1000):
        total = total + run(input_text)
        print("Iteration ", end="")
        print(i)

    if total == -(1 << 63):
        print(total)


def run(input_text: str) -> int:
    lines = split_lines(input_text)
    keymap = MyMap.new()
    key = 0
    for line_obj1 in lines:
        line = line_obj1
        keymap.put(MyStr(line[:3]), key)
        key = key + 1
    keymap.put(MyStr("out"), key)

    svr_key = int(keymap.index_get(MyStr("svr")))
    dac_key = int(keymap.index_get(MyStr("dac")))
    fft_key = int(keymap.index_get(MyStr("fft")))
    out_key = int(keymap.index_get(MyStr("out")))

    graph: list[list[int]] = [[] for _ in range(len(lines) + 1)]
    for line_obj2 in lines:
        line = line_obj2
        key = int(keymap.index_get(MyStr(line[:3])))
        nodes = split_by_space(line[5:])
        node_arr = [0] * len(nodes)
        i = 0
        for node_obj in nodes:
            node_arr[i] = int(keymap.index_get(MyStr(node_obj)))
            i = i + 1
        graph[key] = node_arr
    graph[out_key] = []

    memoize_cache = MyMap.new()

    fft_dac = (
        find_all_paths_dag(graph, memoize_cache, svr_key, fft_key)
        * find_all_paths_dag(graph, memoize_cache, fft_key, dac_key)
        * find_all_paths_dag(graph, memoize_cache, dac_key, out_key)
    )
    dac_fft = (
        find_all_paths_dag(graph, memoize_cache, svr_key, dac_key)
        * find_all_paths_dag(graph, memoize_cache, dac_key, fft_key)
        * find_all_paths_dag(graph, memoize_cache, fft_key, out_key)
    )
    num_paths = fft_dac + dac_fft

    print("RESULT:", end="")
    print(num_paths)
    return 0


def read_stdin() -> str:
    return sys.stdin.read()


def split_lines(input_text: str) -> list[str]:
    return input_text.splitlines()


def split_by_space(value: str) -> list[str]:
    if value == "":
        return []
    return value.split(" ")


def as_hashable(value: object) -> Hashable:
    if not isinstance(value, Hashable):
        raise RuntimeError("Map: key is not Hashable")
    return value


def as_equalable(value: object) -> Equalable:
    if not isinstance(value, Equalable):
        raise RuntimeError("Map: key is not Equalable")
    return value


def index_for(key: object, capacity: int) -> int:
    return as_hashable(key).hash_code() % capacity


if __name__ == "__main__":
    main()

File: aoc/2025/11/part2/solver2.java
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

public final class solver2 {
    private interface Hashable {
        long hashCodeValue();
    }

    private interface Equalable {
        boolean equalsValue(Object other);
    }

    private static final class MyMap {
        private long len = 0L;
        private long capacity;
        private Object[] keys;
        private Object[] values;
        private boolean[] occupied;

        private MyMap(long capacity, Object[] keys, Object[] values, boolean[] occupied) {
            this.capacity = capacity;
            this.keys = keys;
            this.values = values;
            this.occupied = occupied;
        }

        static MyMap newMap() {
            return withCapacity(8L);
        }

        static MyMap withCapacity(long minCapacity) {
            long selectedCapacity = 8L;
            while (selectedCapacity < minCapacity) {
                selectedCapacity = selectedCapacity * 2L;
            }
            int arraySize = toIntExact(selectedCapacity);
            return new MyMap(selectedCapacity, new Object[arraySize], new Object[arraySize], new boolean[arraySize]);
        }

        long len() {
            return this.len;
        }

        boolean contains(Object key) {
            return this.findExistingIndex(key) >= 0;
        }

        Object indexGet(Object key) {
            int index = this.findExistingIndex(key);
            if (index < 0) {
                throw new IllegalStateException("Map.index_get: key not found");
            }
            return this.values[index];
        }

        void put(Object key, Object value) {
            this.maybeGrowForInsert();
            if (this.insertOrAssign(key, value)) {
                this.len = this.len + 1L;
            }
        }

        private void maybeGrowForInsert() {
            if (((this.len + 1L) * 10L) < (this.capacity * 7L)) {
                return;
            }
            this.rehash(this.capacity * 2L);
        }

        private void rehash(long newCapacity) {
            int oldCapacity = toIntExact(this.capacity);
            Object[] oldKeys = this.keys;
            Object[] oldValues = this.values;
            boolean[] oldOccupied = this.occupied;

            this.capacity = newCapacity;
            this.keys = new Object[toIntExact(newCapacity)];
            this.values = new Object[toIntExact(newCapacity)];
            this.occupied = new boolean[toIntExact(newCapacity)];
            this.len = 0L;

            for (int i = 0; i < oldCapacity; i++) {
                if (oldOccupied[i]) {
                    if (this.insertOrAssign(oldKeys[i], oldValues[i])) {
                        this.len = this.len + 1L;
                    }
                }
            }
        }

        private int findExistingIndex(Object key) {
            int capacityI32 = toIntExact(this.capacity);
            int index = indexFor(key, this.capacity);
            int probes = 0;

            while (probes < capacityI32) {
                if (!this.occupied[index]) {
                    return -1;
                }
                if (asEqualable(this.keys[index]).equalsValue(key)) {
                    return index;
                }
                index = index + 1;
                if (index >= capacityI32) {
                    index = 0;
                }
                probes = probes + 1;
            }

            return -1;
        }

        private boolean insertOrAssign(Object key, Object value) {
            int capacityI32 = toIntExact(this.capacity);
            int index = indexFor(key, this.capacity);
            int probes = 0;

            while (probes < capacityI32) {
                if (!this.occupied[index]) {
                    this.occupied[index] = true;
                    this.keys[index] = key;
                    this.values[index] = value;
                    return true;
                }
                if (asEqualable(this.keys[index]).equalsValue(key)) {
                    this.values[index] = value;
                    return false;
                }
                index = index + 1;
                if (index >= capacityI32) {
                    index = 0;
                }
                probes = probes + 1;
            }

            throw new IllegalStateException("Map: table full");
        }
    }

    private static final class MyStr implements Hashable, Equalable {
        final String raw;

        MyStr(String raw) {
            this.raw = raw;
        }

        @Override
        public long hashCodeValue() {
            long hash = 0L;
            for (int i = 0; i < this.raw.length(); i++) {
                hash = hash * 31L + this.raw.charAt(i);
            }
            return hash;
        }

        @Override
        public boolean equalsValue(Object other) {
            if (other == null) {
                return false;
            }
            if (!(other instanceof MyStr rightStr)) {
                return false;
            }
            return this.raw.equals(rightStr.raw);
        }
    }

    private static final class CacheElem implements Hashable, Equalable {
        final long start;
        final long end;

        CacheElem(long start, long end) {
            this.start = start;
            this.end = end;
        }

        @Override
        public boolean equals(Object other) {
            return this.equalsValue(other);
        }

        @Override
        public boolean equalsValue(Object other) {
            if (other == null) {
                return false;
            }
            if (!(other instanceof CacheElem rightCache)) {
                return false;
            }
            return this.start == rightCache.start && this.end == rightCache.end;
        }

        @Override
        public int hashCode() {
            long hash = this.hashCodeValue();
            return (int)(hash ^ (hash >>> 32));
        }

        @Override
        public long hashCodeValue() {
            long hash = 117L;
            hash = hash * 311L + this.start;
            hash = hash * 311L + this.end;
            return hash;
        }
    }

    private solver2() {
    }

    private static long findAllPathsDAG(long[][] graph, MyMap memoizeCache, long start, long end) {
        if (start == end) {
            return 1L;
        }

        CacheElem elem = new CacheElem(start, end);
        if (memoizeCache.contains(elem)) {
            return ((Long)memoizeCache.indexGet(elem)).longValue();
        }

        long paths = 0L;
        for (long node : graph[(int)start]) {
            paths = paths + findAllPathsDAG(graph, memoizeCache, node, end);
        }

        memoizeCache.put(elem, Long.valueOf(paths));
        return paths;
    }

    public static void main(String[] args) throws IOException {
        String input = readStdin();
        long sum = 0L;
        for (long i = 0L; i < 1000L; i++) {
            sum = sum + run(input);
            System.out.print("Iteration ");
            System.out.println(i);
        }

        if (sum == Long.MIN_VALUE) {
            System.out.println(sum);
        }
    }

    private static long run(String input) {
        List<String> lines = splitLines(input);
        MyMap keymap = MyMap.newMap();
        long key = 0L;
        String line;
        for (String lineObj1 : lines) {
            line = lineObj1;
            keymap.put(new MyStr(line.substring(0, 3)), Long.valueOf(key));
            key = key + 1L;
        }
        keymap.put(new MyStr("out"), Long.valueOf(key));

        long svrKey = ((Long)keymap.indexGet(new MyStr("svr"))).longValue();
        long dacKey = ((Long)keymap.indexGet(new MyStr("dac"))).longValue();
        long fftKey = ((Long)keymap.indexGet(new MyStr("fft"))).longValue();
        long outKey = ((Long)keymap.indexGet(new MyStr("out"))).longValue();

        long i = 0L;
        long[][] graph = new long[lines.size() + 1][];
        for (String lineObj2 : lines) {
            line = lineObj2;
            key = ((Long)keymap.indexGet(new MyStr(line.substring(0, 3)))).longValue();
            List<String> nodes = splitBySpace(line.substring(5));
            long[] nodeArr = new long[nodes.size()];
            i = 0L;
            for (String nodeObj : nodes) {
                nodeArr[(int)i] = ((Long)keymap.indexGet(new MyStr(nodeObj))).longValue();
                i = i + 1L;
            }
            graph[(int)key] = nodeArr;
        }
        graph[(int)outKey] = new long[0];

        MyMap memoizeCache = MyMap.newMap();

        long fftDac =
            findAllPathsDAG(graph, memoizeCache, svrKey, fftKey) *
            findAllPathsDAG(graph, memoizeCache, fftKey, dacKey) *
            findAllPathsDAG(graph, memoizeCache, dacKey, outKey);
        long dacFft =
            findAllPathsDAG(graph, memoizeCache, svrKey, dacKey) *
            findAllPathsDAG(graph, memoizeCache, dacKey, fftKey) *
            findAllPathsDAG(graph, memoizeCache, fftKey, outKey);
        long numPaths = fftDac + dacFft;

        System.out.print("RESULT:");
        System.out.println(numPaths);

        return 0L;
    }

    private static String readStdin() throws IOException {
        return new String(System.in.readAllBytes(), StandardCharsets.UTF_8);
    }

    private static List<String> splitLines(String input) {
        ArrayList<String> lines = new ArrayList<>();
        input.lines().forEach(lines::add);
        return lines;
    }

    private static List<String> splitBySpace(String value) {
        if (value.isEmpty()) {
            return new ArrayList<>();
        }
        return new ArrayList<>(List.of(value.split(" ")));
    }

    private static Hashable asHashable(Object value) {
        if (!(value instanceof Hashable hashable)) {
            throw new IllegalStateException("Map: key is not Hashable");
        }
        return hashable;
    }

    private static Equalable asEqualable(Object value) {
        if (!(value instanceof Equalable equalable)) {
            throw new IllegalStateException("Map: key is not Equalable");
        }
        return equalable;
    }

    private static int indexFor(Object key, long capacity) {
        long hash = asHashable(key).hashCodeValue();
        long index = Long.remainderUnsigned(hash, capacity);
        return toIntExact(index);
    }

    private static int toIntExact(long value) {
        if (value < Integer.MIN_VALUE || value > Integer.MAX_VALUE) {
            throw new IllegalStateException("integer overflow");
        }
        return (int)value;
    }
}